File: javascript/leetcode/most-common-word.js
/**
 * @param {string} paragraph
 * @param {string[]} banned
 * @return {string}
 */
var mostCommonWord = function(paragraph, banned) {
    seen = {}
    ban = {}
    banned.forEach(function(w) {
        wKey = w.toLowerCase()
        ban[wKey] = 1
    })
    seenMostCount = 0
    seenMostWord = ''
    paragraph.split(' ').forEach(function(w) {
        wKey = w.replace(/[\?!,\.;'"]/, '').toLowerCase()
        if (!ban[wKey]) {
            if (seen[wKey]) {
                seen[wKey]++
            } else {
                seen[wKey] = 1
            }
            if (seen[wKey] > seenMostCount) {
                seenMostCount = seen[wKey]
                seenMostWord = wKey
            }
        }
    })
    return seenMostWord
};

File: java/leetcode/most-common-word.java
class Solution {
    public String mostCommonWord(String paragraph, String[] banned) {
        HashMap<String, Integer> seen = new HashMap<String, Integer>();
        HashMap<String, Boolean> ban = new HashMap<String, Boolean>();
        String w;
        String wKey;
        for (int i = 0; i < banned.length; i++) {
            w = banned[i];
            wKey = w.toLowerCase();
            ban.put(wKey, true);
        }
        int seenMostCount = 0;
        String seenMostWord = "";
        String[] words = paragraph.split(" ");
        for (int j = 0; j < words.length; j++) {
            w = words[j];
            wKey = w.replaceAll("[\\?!,\\.;'\"]", "").toLowerCase();
            if (!ban.containsKey(wKey)) {
                if (seen.containsKey(wKey)) {
                    seen.put(wKey, seen.get(wKey) + 1);
                } else {
                    seen.put(wKey, 1);
                }
                if (seen.get(wKey) > seenMostCount) {
                    seenMostCount = seen.get(wKey);
                    seenMostWord = wKey;
                }
            }
        }
        return seenMostWord;
    }
}
